src/contentsign/ClickBounty.sol

```diff
@@ -0,0 +1,286 @@
+    }
+
+    // -------------------------------------------------------------------------
+    // Admin Functions
+    // -------------------------------------------------------------------------
+
+    /**
+     * @notice Sets the entry fee for bounties.
+     * @dev Only callable by addresses with DEFAULT_ADMIN_ROLE.
+     * @param _entryFee The new entry fee in units of `token`.
+     */
+    function setEntryFee(uint256 _entryFee) external {
+        _checkRole(DEFAULT_ADMIN_ROLE);
+        entryFee = _entryFee;
+        emit EntryFeeUpdated(_entryFee);
+    }
+
+    /**
+     * @notice Withdraws a specified amount of tokens to a given recipient address.
+     * @dev Only callable by addresses with DEFAULT_ADMIN_ROLE.
+     * @param recipient The address that will receive the tokens.
+     * @param amount    The amount of tokens to withdraw.
+     *
+     * Emits:
+     * - A transfer event from the ERC20 token contract to `recipient`.
+     */
+    function withdraw(address recipient, uint256 amount) external {
+        _checkRole(DEFAULT_ADMIN_ROLE);
+        token.safeTransfer(recipient, amount);
+    }
+
+    // -------------------------------------------------------------------------
+    // Oracle Function
+    // -------------------------------------------------------------------------
+
+    /**
+     * @notice Pays a bounty for a given token ID.
+     *         - If `amount < entryFee`, the difference is collected from the token’s owner.
+     *         - If `amount > entryFee`, the surplus is paid out to the owner.
+     *
+     * @dev Only callable by addresses with ORACLE_ROLE.
+     *
+     * Requirements:
+     * - The token must not have been paid before (`bounties[tokenId] == 0`).
+     *
+     * @param tokenId The token ID to which the bounty is awarded.
+     * @param amount  The bounty amount in `token`.
+     *
+     * Emits:
+     * - A transfer event from the token’s owner to this contract (if `amount < entryFee`).
+     * - A transfer event from this contract to the token’s owner (if `amount > entryFee`).
+     * - A {BountyIssued} event upon successful bounty payment.
+     *
+     * Reverts:
+     * - If `bounties[tokenId] != 0` (bounty already paid).
+     */
+    function awardBounty(uint256 tokenId, uint256 amount) external {
+        _checkRole(ORACLE_ROLE);
+        // Ensure the bounty is non-zero
+        if (amount == 0) {
+            revert ZeroBounty(tokenId);
+        }
+        // Ensure this token ID hasn't been paid yet
+        if (bounties[tokenId] != 0) {
+            revert BountyAlreadyPaid(tokenId);
+        }
+
+        // Identify the token owner via the contentSign contract
+        address owner = contentSign.ownerOf(tokenId);
+
+        // If the bounty is smaller than the entry fee, collect the shortfall from the owner
+        if (amount < entryFee) {
+            token.safeTransferFrom(owner, address(this), entryFee - amount);
+        }
```

Comment:
I don't understand what's the point of this? What is the purpose and can it be made cleaer?

Reply:
Though the user could revoke their allowance before this happens. If so we can't take the tokens from their balance anymore.

I think we should have one function for the user to enter the mission, and one to reward (or slash) them

Reply:
The backend service would check the user's allowance before providing any service. The opportunity window to revoke that allowance is very short (mostly less than 10 seconds before the backend processes a single task). But the benefit for revoking is nothing, because they the user would not receive the service benefits or the results of the backend process if this fails. 
The benefits of having these two transactions combined are
* less delay (UX)
* less failure opportunities (UX)
* less gas fee (Cost)
Having said that, since we can improve for UX and cost later, I'm going to take your suggestion and split them now. We will decide about optimization later if they happen to be important.